fairseq/manifest.py

```diff
@@ -0,0 +1,217 @@
+                    rel_path = os.path.relpath(
+                        self.manifest_audio_paths[i], common_dir
+                    )
+                    rel_path = os.path.join(rel_path, os.path.basename(path))
+                else:
+                    rel_path = os.path.relpath(path, common_dir)
+                if rand.random() > self.valid_percent:
+                    train_data.append((rel_path, frames))
+                else:
+                    print("{}\t{}".format(rel_path, frames), file=valid_f)
+            all_train_data.append(train_data)
+        if valid_f is not None:
+            valid_f.close()
+
+        if self.upsampling_alpha is not None and self.upsampling_alpha < 1:
+            corpora_lengths = []
+            for train_data in all_train_data:
+                corpora_lengths.append(sum([frames for _, frames in train_data]))
+            num_corpora = len(corpora_lengths)
+            sum_corpora_length = sum(corpora_lengths)
+
+            max_length_index = corpora_lengths.index(max(corpora_lengths))
+
+            upsampled_proportions = [
+                (l / sum_corpora_length) ** self.upsampling_alpha
+                for l in corpora_lengths
+            ]
+            upsampled_probabilities = [
+                p / sum(upsampled_proportions) for p in upsampled_proportions
+            ]
+
+            matrix_row_len = np.expand_dims(
+                np.array(
+                    corpora_lengths[:max_length_index]
+                    + corpora_lengths[max_length_index + 1 :],
+                    dtype=float,
+                ),
+                axis=0,
+            )
+            matrix_row_prob = np.expand_dims(
+                np.array(
+                    upsampled_probabilities[:max_length_index]
+                    + upsampled_probabilities[max_length_index + 1 :],
+                    dtype=float,
+                ),
+                axis=0,
+            )
+            if num_corpora > 2:
+                w_mat = np.repeat(matrix_row_len, num_corpora - 1, axis=0)
+                w_mat *= np.transpose(  # element-wise multiplication
+                    np.repeat(matrix_row_prob, num_corpora - 1, axis=0)
+                )
+            else:
+                w_mat = matrix_row_len * matrix_row_prob
+            w_mat -= np.diag(matrix_row_len[0])
+            b_vec = np.array(
+                [
+                    -upsampled_probabilities[i] * corpora_lengths[max_length_index]
+                    for i in range(num_corpora)
+                    if i != max_length_index
+                ]
+            )
+            upsampling_factors = np.linalg.solve(w_mat, b_vec)
+            # assert all entries larger than 1 because we only want to
+            # upsample and never downsample
+            assert min(
+                upsampling_factors >= 1
+            )
+
+            upsampling_factors = np.insert(upsampling_factors, max_length_index, 1.0)
```

Comment:
I finally understood the math here. So in [paper](https://arxiv.org/abs/2006.13979) , it is suggested to upsample the low resource language by `p_l**alpha`, p_l is the calculated by dividing the total number of hours of language l by total number of hours if all languages. For example the language probability is [p1, p2, p3, p4] and alpha is 0.5, after upsampling the probability would be  [p1**0.5, p2**0.5, p3**0.5, p4**0.5], the upsample factor is [p1**0.5/p1, p2**0.5/p2, p3**0.5/p3, p4**0.5/p4]. But here we don't want to upsample the largest language, so the upsample factor is normalised by the minimum. So it would be` min = [p1**0.5/p1, p2**0.5/p2, p3**0.5/p3, p4**0.5/p4]` and final result is` [(p1**0.5/p1)/min, (p2**0.5/p2)/min, (p3**0.5/p3)/min, (p4**0.5/p4)/min]`